### ISSUE TITLE
Train Haloflow on DANN outputs

### ISSUE THREAD

**Issue by Nikhil0504** [galactic-ai/haloflow#23]
Issue to train the output features of the DANN with the haloflow NPEs.

Initially trained a DANN with this config,

```python
sims = ['TNG50', 'TNG100', 'Eagle100', 'Simba100']
config = {
    'obs': 'mags',
    'train_sim': sims[:-1],
    'batch_size': 32,
    'num_epochs': 100,
    'lr': 0.001,
    'feature_layers': [128, 64, 32], 
    'label_layers': [32, 16, 8], 
    'domain_layers': [32, 16, 8], 
    'es_patience': 5,
    'es_min_delta': 0.005
}
```
Model Training outputs: 
```bash
Epoch 100: 100%|██████████| 743/743 [00:01<00:00, 492.35it/s]
Loss task: 0.10282468795776367, Loss domain: 0.7433397173881531
Epoch 100/100
Train Loss: 0.9382, Test Loss: 0.0404, Domain Accuracy: 0.5820
```

![Image](https://github.com/user-attachments/assets/53f3974d-f7b4-4206-aaf5-64d7e1ce00cb)

After training the DANN model, I predicted the outputs in to run NPEs:
https://github.com/galactic-ai/haloflow/blob/8c4f9b173ed90f98b5fbb5280620ac60091ac142/bin/npe/npe_dann.py#L67

and validated in https://github.com/galactic-ai/haloflow/blob/8c4f9b173ed90f98b5fbb5280620ac60091ac142/bin/npe/npe_valid.ipynb
to get:

![Image](https://github.com/user-attachments/assets/90adf4ab-b942-4120-b9fa-b9ff6a69c22e)
![Image](https://github.com/user-attachments/assets/d5776796-2d33-40da-b31f-8d7d85631e4b)

Upon first look, I think this is what we would expect while training and testing with two different simulations but the best way to check would be with the scatter plots of Mh vs M* to get a better idea on what's happening.

**Comment by Nikhil0504** [comment]
@changhoonhahn I did some more tests and used our previous NPE training from #2 to get the coverage plots and scatter plots for the NPE only runs for `SIMBA100` training and testing and the DANN run where we ran with all the other simulations and tested with `SIMBA100` that we have then used for the NPE training.

https://github.com/galactic-ai/haloflow/blob/234f2cd332fdc39d5aeabeb56b6319c76a00e450/tests/npe_dann_scatter_test.ipynb

![Image](https://github.com/user-attachments/assets/df8532ef-117c-4769-886e-2f6fbb449e49)

![Image](https://github.com/user-attachments/assets/e1fbc377-66bc-4955-8cb0-274d6fe81001)

**Comment by changhoonhahn** [comment]
Can you clarify this figure in the nb?
![Image](https://github.com/user-attachments/assets/79c958a8-3a20-4fc6-ba96-816a392d9834)

Is this a comparison of the NPEs trained on Simba100 and the rank statistics also computed on Simba100? If so, the NPE +DANN requires further training. 

I think there might be some confusion with training the NPE+DANN. You want to train NPEs to estimate $p(M_*, M_h | cX_{\rm mag})$ where $cX_{\rm mag}$ is the 2D output of the trained DANN. 

So for training the NPE+DANN for Simba100 you'd 
1. take original training dataset $(Y, X)$
2. compute $cX$ from $X$ by feed it into the trained DANN: $cX = {\rm DANN}(X)$
3. take new training dataset $(Y, cX)$ and split into train/valid/test
4. train bunch of NPE to estimate $p(Y | cX)$ 
5. test best trained NPEs on $(Y_{\rm test}, cX_{\rm test})$ (rank statistics, coverage plot).

**Comment by Nikhil0504** [comment]
> 3. take new training dataset $(Y, cX)$ and split into train/valid/test

Ah yes, that's the bug. I fixed it in:
https://github.com/galactic-ai/haloflow/blob/70a5fcbcc3b17688b11de86f8234d50025691f57/bin/npe/npe_dann.py#L23-L24

Here are the rank statistics with the bug fixed. The blue and green lines are pretty similar to each other so I think the model has been trained pretty well. I also added a cross-correlation model where we train _only_ the NPE with `Eagle100` and test it with `Simba100`. As expected, it wouldn't work as well.

![Image](https://github.com/user-attachments/assets/fd280e1f-c064-4759-80a6-43164ef7189b)

![Image](https://github.com/user-attachments/assets/d0441845-f67c-4b2f-8bcd-9400b07fedbd)

![Image](https://github.com/user-attachments/assets/609c2f9c-5ee9-46bb-b764-f8d0cf9fb7bd)

**Comment by Nikhil0504** [comment]
@changhoonhahn I trained the DANN with `['TNG50', 'TNG100', 'Eagle100']` and tested with `Simba100`. Then I trained the NPEs with the outputs from this DANN model.

Here is the results for `TNG50`. `N` refers for only the NPE runs and `ND` refers to NPE+DANN runs. If we run either `N` or `ND` with the same simulation we get good inference as we expect. But, if we run `TNG50` with `Simba100` we get a curve similar to the run with only the NPE. Does that mean that DANN model didn't do a good job of making a domain agnostic model? **The legend for the coverage plot is wrong for TNG50 (sorry about that)... use the above legend from the rank plot instead.**

![Image](https://github.com/user-attachments/assets/fb6d5375-d9e4-4ec5-8c8c-40377f4d8137)

![Image](https://github.com/user-attachments/assets/dae997a7-c712-4b96-9ceb-2affb1f1f53d)


Here are the plots for `TNG100`. It's closer here but I think this would be the best-case scenario.

![Image](https://github.com/user-attachments/assets/a1f8882b-59e0-441f-b6f7-af85cef964b8)

![Image](https://github.com/user-attachments/assets/6772f37a-0212-4023-8778-dc1de6ee67de)


For `Eagle100`, same problem as `TNG50`

![Image](https://github.com/user-attachments/assets/391abf24-2f76-43a2-b002-e660e53b06ee)

![Image](https://github.com/user-attachments/assets/382bfe7f-f125-46ff-af7e-6de15284ab3e)

**Comment by changhoonhahn** [comment]
Interesting. The coverage plots look like you've been able to train the NPEs successfully. 

I'm a bit confused by how the ND (Eagle100 - Simba100) performs *worse* than N (Eagle100 - Simba100). This also seems to be the case for TNG50. Lets do the following checks: 

1. Let's go back to the DANN. Let's check how the DANN predicted M* and Mh compare against the true values for all the simulations including Simba100. This is to check whether the DANN itself introduces any bias. I'm primarily concerned that the skewed p(M*) and p(Mh) distribution in the training data of DANN cause it to systematically underestimate M* and Mh. 
2. The coverage plots are a good statistical test but we want a more tangible metric of robustness that's more directly translateable to M* and Mh space. Can you make plots of $\Delta M_*$ versus true $M_*$ and $\Delta M_h$ versus true $M_h$? That would directly reveal the bias and it's amplitude.

**Comment by Nikhil0504** [comment]
@changhoonhahn I finally got the DANN working! See commit above for the code. 

I ran the DANN models for the all the simulations for the `mags` obs. 
Each plot is trained with 3 sims and tested with the 4th sim.
`Sims -> Train Sim with obs mag`

![Image](https://github.com/user-attachments/assets/4bfdafff-2745-4e9e-a70a-0947673894ff)
![Image](https://github.com/user-attachments/assets/da61f3b2-3d34-40f4-a457-863bac6b31ca)
![Image](https://github.com/user-attachments/assets/de19196a-5618-45e8-a48f-b11efd1f9918)
![Image](https://github.com/user-attachments/assets/2859a49e-c5b4-46f4-be98-8b7e69d01d95)

**Comment by changhoonhahn** [comment]
Hm, I looked through your code `bin/dann/train.py` and I think there's still a bug in how you whiten the data. 

In the training, 
https://github.com/galactic-ai/haloflow/blob/5abe998e720905d9af6b7c30ae30f415072abc65/bin/dann/train_new.py#L32-L56
you whiten using the mean and std from *all* X excluding `sim`.

Then in eval, 
https://github.com/galactic-ai/haloflow/blob/5abe998e720905d9af6b7c30ae30f415072abc65/src/haloflow/dann/evalutate.py#L12
you whiten using the mean and std from `X_eval` only. This should be whitened in the exactly same way.

**Comment by Nikhil0504** [comment]
Fixed in afb3985ee3bd7e4db5ad6ce877efef211cfa25e2

![Image](https://github.com/user-attachments/assets/f3df19ed-69ee-476e-bd2e-f3587a6d04f8)

![Image](https://github.com/user-attachments/assets/03b82563-2544-4241-a340-210d204c5fec)

![Image](https://github.com/user-attachments/assets/082dfdfb-3677-4e42-b9e6-b2ca3e415b44)

![Image](https://github.com/user-attachments/assets/6b5532de-45a0-4276-835b-d44a7fba410a)

**Comment by Nikhil0504** [comment]
@changhoonhahn the model does work better with `mags_morph_extra` but I think we will need to hyperparameter tuning next before going to the NPEs.

![Image](https://github.com/user-attachments/assets/1029dc03-bce9-48cb-a73c-1aea0b985367)

![Image](https://github.com/user-attachments/assets/0180c94a-d4a2-45ca-a0ce-8ec01c769ab2)

**Comment by Nikhil0504** [comment]
@changhoonhahn I think they dynamic range of the combined `schechter_logmass` and `count_weights` is too high for the MSE loss function to work well. So I took the negative log of it to get the range of the weights down. What do you think I could do to improve it further?

`weights = -np.log10(count_weights * sche_weights) + 1e-8`
Min weight: 1.8701595092346435
Max weight: 6.749358865899538
![Image](https://github.com/user-attachments/assets/698d3799-723c-4790-863e-946de2beab08)

`weights = count_weights * sche_weights`
Min weight: 1.7809066028715122e-07
Max weight: 0.013484675526910638
![Image](https://github.com/user-attachments/assets/320b44f6-3a0b-48ee-95d4-8dca88e31a9e)

`weights = count_weights`
![Image](https://github.com/user-attachments/assets/68ff7f64-2110-4979-832e-a0eb12d8e93c)

`weights = None`
![Image](https://github.com/user-attachments/assets/0f60e3d7-0e20-4e07-8b61-8690051fc217)

**Comment by Nikhil0504** [comment]
Fixed the bug in https://github.com/galactic-ai/haloflow/blob/ff182c070770ed85694fffaef28aa763786c0c34/bin/dann/train_new.py#L57-L65

![Image](https://github.com/user-attachments/assets/001eadb4-1b16-4942-a24d-9398904c86f8)

![Image](https://github.com/user-attachments/assets/1b5e2c76-cd89-418a-b5fb-1ff1b7f3495f)

![Image](https://github.com/user-attachments/assets/d8604b2b-de49-45fa-bcb2-3c6fa83fa7f5)
